src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1381 @@
+                // - if we are not negating, we can expand it to `[_, _, ...] and [L1, _, ...] and [_, L2, ...] and ...`
+                //   and the `and` and `or` patterns in the element patterns can then be lifted out further.
+                //
+                // - if we are negating, we can expand it to `not [_, _, ...] or [not L1, _, ...] or [_, not L2, ...] or ...`
+                //   and the `and` and `or` patterns in the resulting element patterns can then be lifted out further.
+
+                var saveExpectingOperandOfDisjunction = _expectingOperandOfDisjunction;
+                int startOfLeft = _evalSequence.Count;
+
+                // All the operands we push from here will be combined in `or` for negation and `and` otherwise
+                _expectingOperandOfDisjunction = _negated;
+
+                // `[_, _, ..., .._]` (effectively a not null and Length test)
+                ImmutableArray<BoundPattern> equivalentDefaultPatterns = listPattern.Subpatterns.SelectAsArray(makeEquivalentDefaultPattern);
+                BoundListPattern lengthTest = listPattern.WithSubpatterns(equivalentDefaultPatterns);
+                TryPushOperand(NegateIfNeeded(lengthTest));
+                Debug.Assert(_evalSequence.Count == startOfLeft + 1);
+
+                int startOfNestedPatterns = _evalSequence.Count;
+                var saveMakeEvaluationSequenceOperand = _makeEvaluationSequenceOperand;
+
+                int i = 0;
+
+                // Given `newPattern`, produce `[..., _, newPattern, _, ...]`
+                Func<BoundPattern, BoundPattern> makeListPattern = (BoundPattern newPattern) =>
+                {
+                    // Note: lambda intentionally captures
+                    bool wasCompilerGenerated = newPattern.WasCompilerGenerated;
+                    newPattern = WithInputTypeCheckIfNeeded(newPattern, equivalentDefaultPatterns[i].InputType);
+                    ImmutableArray<BoundPattern> newSubpatterns = equivalentDefaultPatterns.SetItem(i, newPattern);
+
+                    BoundPattern newList = new BoundListPattern(
+                        newPattern.Syntax, newSubpatterns, hasSlice: newSubpatterns.Any(p => p is BoundSlicePattern), listPattern.LengthAccess, listPattern.IndexerAccess,
+                        listPattern.ReceiverPlaceholder, listPattern.ArgumentPlaceholder, listPattern.Variable, listPattern.VariableAccess,
+                        listPattern.InputType, listPattern.NarrowedType);
+
+                    if (wasCompilerGenerated)
+                    {
+                        newList = newList.MakeCompilerGenerated();
+                    }
+
+                    return saveMakeEvaluationSequenceOperand?.Invoke(newList) ?? newList;
+                };
+
+                // Given `newPattern`, produce `[..., _, ..newPattern, _, ...]`
+                Func<BoundPattern, BoundPattern> makeListPatternWithSlice = (BoundPattern newPattern) =>
```

Comment:
>makeListPatternWithSlice [](http://example.com/codeflow?start=49&length=24)

It looks like we can predict whether any of the new delegates will be used by checking information in ```listPattern``` and allocate delegates conditionally.